traitsui/ui_editors/data_frame_editor.py

```diff
@@ -0,0 +1,309 @@
+from traitsui.toolkit import toolkit_object
+from traitsui.qt4.ui_editor import UIEditor
+from traitsui.view import View
+
+
+class DataFrameAdapter(TabularAdapter):
+    """ Generic tabular adapter for data frames
+    """
+
+    #: The text to use for a generic entry.
+    text = Property
+
+    #: The alignment for each cell
+    alignment = Property(Enum('left', 'center', 'right'))
+
+    #: The text to use for a row index.
+    index_text = Property
+
+    #: The alignment to use for a row index.
+    index_alignment = Property
+
+    #: The font to use for each column
+    font = Property
+
+    #: The format to use for each column
+    format = Property
+
+    #: The format for each element, or a mapping column ID to format.
+    _formats = Either(Str, Dict, default='%s')
+
+    #: The font for each element, or a mapping column ID to font.
+    _fonts = Either(Font, Dict, default='Courier 10')
+
+    def _get_index_alignment(self):
+        import numpy as np
+
+        index = getattr(self.object, self.name).index
+        if np.issubdtype(index.dtype, np.number):
+            return 'right'
+        else:
+            return 'left'
+
+    def _get_alignment(self):
+        import numpy as np
+
+        column = self.item[self.column_id]
+        if np.issubdtype(column.dtype, np.number):
+            return 'right'
+        else:
+            return 'left'
+
+    def _get_font(self):
+        if isinstance(self._fonts, toolkit_object('font_trait:TraitsFont')):
+            return self._fonts
+        else:
+            return self._fonts.get(self.column_id, 'Courier 10')
+
+    def _get_format(self):
+        if isinstance(self._formats, basestring):
+            return self._formats
+        else:
+            return self._formats.get(self.column_id, '%s')
+
+    def _get_content(self):
+        return self.item[self.column_id].iloc[0]
+
+    def _get_text(self):
+        format = self.get_format(self.object, self.name, self.row, self.column)
+        return format % self.get_content(self.object, self.name, self.row,
+                                         self.column)
+
+    def _set_text(self, value):
+        column = self.item[self.column_id]
+        dtype = column.dtype
+        value = dtype.type(value)
+        column.iloc[0] = value
+
+    def _get_index_text(self):
+        return str(self.item.index[0])
+
+    def _set_index_text(self, value):
+        index = getattr(self.object, self.name).index
+        dtype = index.dtype
+        value = dtype.type(value)
+        index.values[self.row] = value
+
+    #---- Adapter methods that are not sensitive to item type ----------------
+
+    def get_item(self, object, trait, row):
+        """ Override the base implementation to work with DataFrames
+
+        This returns a dataframe with one row, rather than a series, since
+        using a dataframe preserves dtypes.
+
+        """
+        return getattr(object, trait).iloc[row:row+1]
```

Comment:
This fails with Pandas 0.12 with an `IndexError` when the dataframe being viewed has no rows. Newer versions of Pandas are fine.